CRM/PCP/BAO/PCP.php

```diff
@@ -931,5 +931,286 @@ public static function getOwnerNotificationId($component_id, $component = 'contr
+        LEFT JOIN civicrm_email supporter_email ON (pcp_supporter.id = supporter_email.contact_id AND supporter_email.is_primary = 1)
+        WHERE $whereClause
+        GROUP BY pcps.id
+        {$orderBy}
+        {$limit}";
+        
+    $object = CRM_Core_DAO::executeQuery($query, $params, TRUE, 'CRM_PCP_DAO_PCP');
+
+    while ($object->fetch()) {
+      $links = self::actionLinks();
+      
+      $approvedId = CRM_Core_OptionGroup::getValue('pcp_status', 'Approved', 'name');
+      $allowToDelete = CRM_Core_Permission::check('delete in CiviContribute');
+
+      CRM_Core_DAO::storeValues($object, $values[$object->id]);
+      
+      if ($object->page_type == 'contribute') {
+        $pageUrl = CRM_Utils_System::url('civicrm/' . $object->page_type . '/transact', 'reset=1&id=' . $object->page_id);
+      }
+      else {
+        $pageUrl = CRM_Utils_System::url('civicrm/' . $object->page_type . '/register', 'reset=1&id=' . $object->page_id);
+      }
+      $supporterUrl = CRM_Utils_System::url('civicrm/contact/view', "reset=1&cid={$object->contact_id}");
+      $pcpUrl = CRM_Utils_System::url('civicrm/pcp/info', "reset=1&id={$object->id}");
+      $status = CRM_PCP_BAO_PCP::buildOptions('status_id', 'create');
+      $values[$object->id]['pcp_title'] = "<a href='{$pcpUrl}'>{$object->title}</a>";
+      $values[$object->id]['pcp_supporter'] = "<a href='{$supporterUrl}'>{$object->display_name}</a>";
+      $values[$object->id]['pcp_page'] = "<a href='{$pageUrl}'>{$object->page}</a>";
+      $values[$object->id]['pcp_start_date'] = CRM_Utils_Date::customFormat($object->start_date);
+      $values[$object->id]['pcp_end_date'] = CRM_Utils_Date::customFormat($object->end_date);
+      $values[$object->id]['pcp_status'] = $status[$object->status_id];
+      
+      $action = array_sum(array_keys($links));
+
+      $class = '';
+
+      if ($object->status_id != $approvedId || $object->is_active != 1) {
+        $class = 'disabled';
+      }
+
+      switch ($object->status_id) {
+        case 2:
+          $action -= CRM_Core_Action::RENEW;
+          break;
+
+        case 3:
+          $action -= CRM_Core_Action::REVERT;
+          break;
+      }
+
+      switch ($object->is_active) {
+        case 1:
+          $action -= CRM_Core_Action::ENABLE;
+          break;
+
+        case 0:
+          $values[$object->id]['class'][] = 'disabled';
+          $action -= CRM_Core_Action::DISABLE;
+          break;
+      }
+
+      if (!$allowToDelete) {
+        $action -= CRM_Core_Action::DELETE;
+      }
+      
+      $values[$object->id]['action'] = CRM_Core_Action::formLink($links,
+          $action,
+          array('id' => $object->id), ts('more'), FALSE, 'pcp.selector.row', 'PCP', $object->id
+        );
+
+    }
+    return $values;
+  }
+  
+   /**
+   * @param array $params
+   *
+   * @return NULL|string
+   */
+  public static function getPcpCount(&$params) {
```

Comment:
We have loads of this passing as reference in the code base- but it makes for fragile code because the $params array gets altered unintentionally & then downstream someone hangs functionality off that change that breaks when someone does a tidy up.